### ISSUE TITLE
[HOLD for Payment 2024-08-22][$250] Chat - Message marker disappears when the 2nd user deletes the first message

### ISSUE THREAD

**Issue by lanitochka17** [Expensify/App#41935]
If you haven’t already, check out our [contributing guidelines](https://github.com/Expensify/ReactNativeChat/blob/main/contributingGuides/CONTRIBUTING.md) for onboarding and email contributors@expensify.com to request to join our Slack channel!
___

**Version Number:** 1.4.72-0
**Reproducible in staging?:** Y
**Reproducible in production?:** Y
**If this was caught during regression testing, add the test name, ID and link from TestRail:** https://expensify.testrail.io/index.php?/tests/view/4550091
**Email or phone of affected tester (no customers):** esraa55saad+123@gmail.com
**Issue reported by:** Applause - Internal Team

## Action Performed:
Precondition: User A in main testing device, user B in secondary device
1. Open the testing app as first user
2. Open any chat as user A (not the chat with user B)
3. As a use B send multiple messages (e..g, 1,2,3,4)
4. As user A open the chat with user B and observe the new message marker
5. As user B delete the first message sent (under the new message marker)

## Expected Result:
New message marker to change place to the 2nd new message sent

## Actual Result:
New message marker disappears

## Workaround:
Unknown

## Platforms:

Which of our officially supported platforms is this issue occurring on?
- [ ] Android: Native
- [x] Android: mWeb Chrome
- [ ] iOS: Native
- [ ] iOS: mWeb Safari
- [x] MacOS: Chrome / Safari
- [ ] MacOS: Desktop

## Screenshots/Videos

Add any screenshot/video evidence

</details> 

https://github.com/Expensify/App/assets/78819774/44d3f38d-ba08-44c6-9664-d30f8bf37890

[View all open jobs on GitHub](https://github.com/Expensify/App/issues?q=is%3Aopen+is%3Aissue+label%3A%22Help+Wanted%22)

<details><summary>Upwork Automation - Do Not Edit</summary>
    <ul>
        <li>Upwork Job URL: https://www.upwork.com/jobs/~01074a670eb87f437f</li>
        <li>Upwork Job ID: 1790122767878377472</li>
        <li>Last Price Increase: 2024-05-20</li>
    </ul>
</details>

<details><summary>Issue Owner</summary>Current Issue Owner: @arosiclair</details>

**Comment by lanitochka17** [comment]
@adelekennedy  FYI I haven't added the External label as I wasn't 100% sure about this issue. Please take a look and add the label if you agree it's a bug and can be handled by external contributors

**Comment by lanitochka17** [comment]
We think that this bug might be related to #vip-vsp

**Comment by adelekennedy** [comment]
still can't reproduce but I think this is due to site slowness

**Comment by adelekennedy** [comment]
aha - finally able to reproduce on chrome

**Comment by adelekennedy** [comment]
@allroundexperts one proposal to review above

**Comment by allroundexperts** [comment]
On it!

**Comment by tsa321** [comment]
## Proposal

### Please re-state the problem that we are trying to solve in this issue.

Unread message marker disappears when the other user delete first marked unread message of the current user.

### What is the root cause of that problem?

When other user delete a message,`calculateUnreadMarker` will be called in [here](https://github.com/Expensify/App/blob/a5bd1f2f7941fd65a5cbf647ac64691cd51b7bc6/src/pages/home/report/ReportActionsList.tsx#L486). This function mainly calling `shouldDisplayNewMarker` in [here](https://github.com/Expensify/App/blob/a5bd1f2f7941fd65a5cbf647ac64691cd51b7bc6/src/pages/home/report/ReportActionsList.tsx#L442) to determine whether marker should be displayed. In shouldDisplayNewMarker because currentUnreadMarker is exist then will go in here: 

https://github.com/Expensify/App/blob/a5bd1f2f7941fd65a5cbf647ac64691cd51b7bc6/src/pages/home/report/ReportActionsList.tsx#L460

But because of the deleted report action the report action will not be include in sortedVisibleReportActions, the marker would not be found:

https://github.com/Expensify/App/blob/a5bd1f2f7941fd65a5cbf647ac64691cd51b7bc6/src/pages/home/report/ReportActionsList.tsx#L492-L494

### What changes do you think we should make in order to solve the problem?

We could add useRef to store next report action id of current marked report action, for example
`const reportActionIdNextToCurrentMarker = useRef(null)`

Then below this line:
https://github.com/Expensify/App/blob/a5bd1f2f7941fd65a5cbf647ac64691cd51b7bc6/src/pages/home/report/ReportActionsList.tsx#L461

We set the ref:
``` javascript
if (shouldDisplay) {
    reportActionIdNextToCurrentMarker.current = nextMessage.reportActionID;
}
```
We must move the nextMessage declaration outside of the if clause first.
<br />

Then above this [line](https://github.com/Expensify/App/blob/a5bd1f2f7941fd65a5cbf647ac64691cd51b7bc6/src/pages/home/report/ReportActionsList.tsx#L490) , we check whether the currentUnreadMarker exist in sortedVisibleReportActions and change the currentUnreadMarker if it is not exist. The code could be:

```javascript
if (currentUnreadMarker) {
    const isCurrentMarkerReportActionExist = sortedVisibleReportActions?.findIndex((reportAction) => {
        return reportAction.reportActionID === currentUnreadMarker;
    });
    if (isCurrentMarkerReportActionExist < 0) {
        console.log(JSON.stringify(reportActionIdNextToCurrentMarker), " lalalalala ", report.reportID);
        setCurrentUnreadMarker(reportActionIdNextToCurrentMarker.current);
        return;
    }
} else {
    reportActionIdNextToCurrentMarker.current = null;
}
```

<details><summary>Result: </summary>

https://github.com/Expensify/App/assets/114975543/352d93e8-0031-49a7-8dff-e5b0a9cb99a2


</details>

**Comment by tsa321** [comment]
@dominictb How to delete multiple message at once? I can only delete one message at a time?

**Comment by adelekennedy** [comment]
@allroundexperts will review the proposals above (chill out Melvin)

**Comment by allroundexperts** [comment]
@dominictb By reading your RCA:
```
Since the user is inside the report, at this point the lastReadTimeRef.current is already set to latest report lastReadTime (it will be set as such after the initial unread marker is calculated), so no report action is considered "unread" now, so the New marker disappears.
```
It seems to me that when a message is deleted, the `report.lastReadTime` gets reset from the backend. Given this, I think a fix from the backend might be a better idea. Let me know if I've understood this correctly.

**Comment by allroundexperts** [comment]
@dominictb Can you reply on [this](https://github.com/Expensify/App/issues/41935#issuecomment-2121804025)? TIA!

**Comment by allroundexperts** [comment]
Ah got it. Thanks for the explanation.

@dominictb's [proposal](https://github.com/Expensify/App/issues/41935#issuecomment-2109618452) makes sense and looks good to me. Let's go with them since I agree that @tsa321's proposal is essentially sharing the same idea as what @dominictb proposed earlier.

🎀 👀 🎀 C+ reviewed

**Comment by arosiclair** [comment]
> Since the user is inside the report, at this point the lastReadTimeRef.current is already set to latest report lastReadTime (it will be set as such after the initial unread marker is calculated), so no report action is considered "unread" now, so the New marker disappears.

This does not sound like much of a bug to me, but maybe I don't fully understand. When exactly is the unread marker supposed to disappear?

**Comment by tsa321** [comment]
@arosiclair how about my [solution](https://github.com/Expensify/App/issues/41935#issuecomment-2119069082)?  I am only need useRef for `currentUnreadMarker`: 
 `const reportActionIdNextToCurrentMarker = useRef(null)`, since user cannot delete multiple messages at once.
 
But if some how user can delete multiple message at once, we can store useRef of report actions ids after current unread marker.

and My RCA is a little bit different. 
The determination to display unread marker is based on `setCurrentMarker`. Even if lastReadTime.current is updated,  the setCurrentMarker to reportActionID will make unread marker visible if the report action exist in sortedVisibleReportAction and setCurrentMarker to  null will hide this marker.

**Comment by arosiclair** [comment]
> https://github.com/Expensify/App/issues/41935#issuecomment-2109618452 to add it as the main solution (original solution moved to alternative section), let me know if it looks good!

Looks like it's in the right direction. We'll refine it in the pull request.

**Comment by arosiclair** [comment]
There was a regression and the PR for this was [reverted](https://github.com/Expensify/App/pull/44247) @allroundexperts @dominictb. Can you fix the regression and re-raise the PR?

**Comment by adelekennedy** [comment]
@dominictb will you have an update available this week?

**Comment by arosiclair** [comment]
That makes sense to me 👍

**Comment by arosiclair** [comment]
Not overdue @MelvinBot

**Comment by arosiclair** [comment]
Original PR was reverted. We're still waiting on the follow up @dominictb

**Comment by arosiclair** [comment]
Per my [last comment on the PR](https://github.com/Expensify/App/pull/44856#issuecomment-2228975746), I'm going to take this internal. I'll try taking the approach I outlined on [this other issue](https://github.com/Expensify/App/issues/44088).

**Comment by arosiclair** [comment]
> @arosiclair following up on your comment here:[#44856 (comment)](https://github.com/Expensify/App/pull/44856#issuecomment-2228975746)
> 
> Shall we proceed the payment here (half the original bounty due to regression)? Thank you!

Unfortunately there will be no payment since your original PR had a regression and was reverted and we did not accept your follow up PR (so ultimately there was nothing delivered).

**Comment by arosiclair** [comment]
I believe they just made an exception in that case it's not a rule. I'll ask internally if we should do that again.

**Comment by mallenexpensify** [comment]
@dominictb this issues is different than https://github.com/Expensify/App/issues/41197#issuecomment-2165639206 because we gave you another chance to submit an acceptable PR.  Since the follow up PR was overly complex and deviated from their proposal, @arosiclair chose to take this internal.  

If the PR that ends up fixing the bug contains code from your followup PR, then you might be due compensation.  We can address that once the PR hits production.  Thx

**Comment by arosiclair** [comment]
Draft is done I just need to complete the checklist and post the PR

**Comment by arosiclair** [comment]
^ this regression got fixed in https://github.com/Expensify/App/pull/47393, but the "fix" is really just a partial revert. So I still have to go back and clean it up.

**Comment by ikevin127** [comment]
⚠️ Production deploy automation failed here -> this should be on **[HOLD for Payment 2024-08-22]** according to yesterday’s production deploy confirmed in https://github.com/Expensify/App/pull/45797#issuecomment-2294129156 and [deploy checklist](https://github.com/Expensify/App/issues/47356).

I reviewed the PR here as C+.

cc @adelekennedy

**Comment by ikevin127** [comment]
cc @adelekennedy

**Comment by adelekennedy** [comment]
payment due today!

I think the only payment due here is for @ikevin127 for reviewing the PR, I've asked internally about any other additional payments.


**Payments Due**
- [ ] Contributor+: $250 @ikevin127 

Upwork offer is [here](https://www.upwork.com/nx/wm/offer/103644958).

**Comment by ikevin127** [comment]
@adelekennedy Offer accepted, thank you!
